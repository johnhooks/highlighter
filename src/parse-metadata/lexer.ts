type PairStart = "{" | "(" | '"' | "'" | "/";
type PairEnd = "}" | ")" | '"' | "'" | "/";

export type Token = {
  value: string;
  start: number;
  end: number;
  type: "identifier" | "literal" | "symbol";
};

type State = "identifier" | "literal" | null;

/**
 * Lexer for Markdown code fence meta data parsing.
 *
 * @internal
 */
export function lexer(input: string): readonly Token[] {
  const tokens: Token[] = [];

  let pos = 0;
  let pair: PairEnd | null = null;
  let start = 0;
  let state: State = null;
  let accumulator: string[] = [];

  loop: while (pos <= input.length) {
    const char = input[pos];

    if (state === "identifier") {
      if (isIdentifier(char)) {
        accumulator.push(char);
      } else {
        tokens.push({ type: "identifier", value: accumulator.join(""), start, end: pos });
        accumulator = [];
        state = null;
      }
    } else if (state === "literal") {
      if (isPairEnd(char, pair)) {
        if (accumulator.length > 0) {
          tokens.push({ type: "literal", value: accumulator.join(""), start, end: pos });
        }
        tokens.push({ type: "symbol", value: char, start: pos, end: pos + 1 });
        accumulator = [];
        state = null;
        // The closing symbol was just lexed, need to break out before `isPairStart`
        pos = pos + 1;
        continue loop;
      } else {
        // Safety to break out of literal if no more `input`
        if (char === undefined) {
          tokens.push({ type: "literal", value: accumulator.join(""), start, end: pos });
          break loop;
        } else {
          accumulator.push(char);
        }
      }
    }

    // Continue lexing `identifier` or `literal`
    if (state !== null) {
      pos = pos + 1;
      continue loop;
    }

    if (isPairStart(char)) {
      switch (char) {
        case "{":
          pair = "}";
          break;
        case "(":
          pair = ")";
          break;
        default:
          pair = char;
      }
      start = pos + 1;
      state = "literal";
      tokens.push({ type: "symbol", value: char, start: pos, end: pos + 1 });
    } else if (isSymbol(char)) {
      tokens.push({ type: "symbol", value: char, start: pos, end: pos + 1 });
    } else if (isIdentifier(char)) {
      start = pos;
      state = "identifier";
      accumulator = [char];
    }

    // Always advance.
    pos = pos + 1;
  }

  return tokens;
}

function isPairEnd(char: unknown, pairEnd: PairEnd | null): boolean {
  return typeof char === "string" && char === pairEnd;
}

function isPairStart(char: unknown): char is PairStart {
  return typeof char === "string" && /^[{("'/]/.test(char);
}

function isIdentifier(char: unknown): boolean {
  return typeof char === "string" && /^[A-Za-z]/.test(char);
}

function isSymbol(char: unknown): boolean {
  return typeof char === "string" && /^[=]/.test(char);
}
