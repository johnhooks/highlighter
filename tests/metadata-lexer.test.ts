import { lexer } from "../src/parse-metadata/lexer.js";

describe("lexer", () => {
  describe("tokens", () => {
    const tokens = lexer('title="testing lexer"');
    it("should lex an identifier token", () => {
      const [identifier] = tokens;
      expect(identifier).toEqual({ value: "title", type: "identifier", start: 0, end: 5 });
    });

    it("should lex symbol tokens", () => {
      const [, equal, leftQuote, , rightQuote] = tokens;
      expect(equal).toEqual({ value: "=", type: "symbol", start: 5, end: 6 });
      expect(leftQuote).toEqual({ value: '"', type: "symbol", start: 6, end: 7 });
      expect(rightQuote).toEqual({ value: '"', type: "symbol", start: 20, end: 21 });
    });

    it("should lex a literal token", () => {
      const [, , , literal] = tokens;
      expect(literal).toEqual({ value: "testing lexer", type: "literal", start: 7, end: 20 });
    });
  });

  describe("snapshots", () => {
    it("should lex a single identifier", () => {
      const tokens = lexer(`javascript`);
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a single quote pair", () => {
      const tokens = lexer("'single'");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a double quote pair", () => {
      const tokens = lexer('"double"');
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a dash pair", () => {
      const tokens = lexer("/word/");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a paren pair", () => {
      const tokens = lexer("(argument)");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a brace pair", () => {
      const tokens = lexer("{range}");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a title", () => {
      const tokens = lexer('title="hello lexer"');
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a range", () => {
      const tokens = lexer("{3..5,7}");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a `showLineNumbers `identifier", () => {
      const tokens = lexer("showLineNumbers");
      expect(tokens).toMatchSnapshot();
    });

    it("should lex a starting line number", () => {
      const tokens = lexer("showLineNumbers{5}");
      expect(tokens).toMatchSnapshot();
    });
  });
});
