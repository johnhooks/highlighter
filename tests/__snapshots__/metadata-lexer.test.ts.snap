// Vitest Snapshot v1

exports[`lexer > snapshots > should lex a \`showLineNumbers \`identifier 1`] = `
[
  {
    "end": 15,
    "start": 0,
    "type": "identifier",
    "value": "showLineNumbers",
  },
]
`;

exports[`lexer > snapshots > should lex a brace pair 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "{",
  },
  {
    "end": 6,
    "start": 1,
    "type": "literal",
    "value": "range",
  },
  {
    "end": 7,
    "start": 6,
    "type": "symbol",
    "value": "}",
  },
]
`;

exports[`lexer > snapshots > should lex a dash pair 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "/",
  },
  {
    "end": 5,
    "start": 1,
    "type": "literal",
    "value": "word",
  },
  {
    "end": 6,
    "start": 5,
    "type": "symbol",
    "value": "/",
  },
]
`;

exports[`lexer > snapshots > should lex a double quote pair 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "\\"",
  },
  {
    "end": 7,
    "start": 1,
    "type": "literal",
    "value": "double",
  },
  {
    "end": 8,
    "start": 7,
    "type": "symbol",
    "value": "\\"",
  },
]
`;

exports[`lexer > snapshots > should lex a paren pair 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "(",
  },
  {
    "end": 9,
    "start": 1,
    "type": "literal",
    "value": "argument",
  },
  {
    "end": 10,
    "start": 9,
    "type": "symbol",
    "value": ")",
  },
]
`;

exports[`lexer > snapshots > should lex a range 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "{",
  },
  {
    "end": 7,
    "start": 1,
    "type": "literal",
    "value": "3..5,7",
  },
  {
    "end": 8,
    "start": 7,
    "type": "symbol",
    "value": "}",
  },
]
`;

exports[`lexer > snapshots > should lex a single identifier 1`] = `
[
  {
    "end": 10,
    "start": 0,
    "type": "identifier",
    "value": "javascript",
  },
]
`;

exports[`lexer > snapshots > should lex a single quote pair 1`] = `
[
  {
    "end": 1,
    "start": 0,
    "type": "symbol",
    "value": "'",
  },
  {
    "end": 7,
    "start": 1,
    "type": "literal",
    "value": "single",
  },
  {
    "end": 8,
    "start": 7,
    "type": "symbol",
    "value": "'",
  },
]
`;

exports[`lexer > snapshots > should lex a starting line number 1`] = `
[
  {
    "end": 15,
    "start": 0,
    "type": "identifier",
    "value": "showLineNumbers",
  },
  {
    "end": 16,
    "start": 15,
    "type": "symbol",
    "value": "{",
  },
  {
    "end": 17,
    "start": 16,
    "type": "literal",
    "value": "5",
  },
  {
    "end": 18,
    "start": 17,
    "type": "symbol",
    "value": "}",
  },
]
`;

exports[`lexer > snapshots > should lex a title 1`] = `
[
  {
    "end": 5,
    "start": 0,
    "type": "identifier",
    "value": "title",
  },
  {
    "end": 6,
    "start": 5,
    "type": "symbol",
    "value": "=",
  },
  {
    "end": 7,
    "start": 6,
    "type": "symbol",
    "value": "\\"",
  },
  {
    "end": 18,
    "start": 7,
    "type": "literal",
    "value": "hello lexer",
  },
  {
    "end": 19,
    "start": 18,
    "type": "symbol",
    "value": "\\"",
  },
]
`;
